### PR TITLE
Keycloak config for MIT Open

### PR DESCRIPTION
**Pull request by shaidar**
# What are the relevant tickets?

https://github.com/mitodl/mit-open/issues/21
# Description (What does it do?)

This is an initial pass to setup the following:
- New realm
- Enables Google reCAPTHCA v2 as part of the registration flow
- Sets up a password policy
- Basic setup to configure Touchstone as an IDP. Still missing some configs for it and need to do some more digging and reach out to IST for a bit more information/help on configuration.

# How can this be tested?

Run it on QA and test out the added features listed above.

### CODE REVIEW

File: src/ol_infrastructure/substructure/keycloak/__main__.py [hunk 1]
```diff
@@ -145,3 +148,140 @@
...
+
+# Create OL Public Realm
+ol_apps_realm = keycloak.Realm(
+    "olapps",
+    access_code_lifespan="30m",
+    access_code_lifespan_user_action="15m",
+    attributes={
+        "business_unit": f"operations-{env_name}",
+    },
+    display_name="OLApps",
```
Comment (blarghmatey) [inline src/ol_infrastructure/substructure/keycloak/__main__.py:160]:
```suggestion
    display_name="OL Apps",
```

File: src/ol_infrastructure/substructure/keycloak/__main__.py [hunk 2]
```diff
@@ -145,3 +148,140 @@
...
+    display_name_html="<b>MIT Open Learning Apps</b>",
```
Comment (blarghmatey) [inline src/ol_infrastructure/substructure/keycloak/__main__.py:161]:
```suggestion
    display_name_html="<b>MIT Open Learning Applications</b>",
```

File: src/ol_infrastructure/substructure/keycloak/__main__.py [hunk 3]
```diff
@@ -145,3 +148,140 @@
...
+        digits=6,
+        initial_counter=2,
+        look_ahead_window=1,
+        period=30,
+        type="totp",
+    ),
+    realm="olapps",
+    reset_password_allowed=True,
+    verify_email=True,
+    password_policy="upperCase(2) and digits(2) and length(12) and specialChars(2) and forceExpiredPasswordChange(365) and notUsername and notEmail",  # noqa: E501,S106 # pragma: allowlist secret
```
Comment (blarghmatey) [inline src/ol_infrastructure/substructure/keycloak/__main__.py:176]: I'm not sure that we want to force end-users to rotate their passwords anually. Maybe also change the requirements for uppercase/digits/special characters to 1? And maybe require at least 2 of the 3?

File: src/ol_infrastructure/substructure/keycloak/__main__.py [hunk 4]
```diff
@@ -145,3 +148,140 @@
...
+        starttls=True,
+    ),
+    ssl_required="external",
+    offline_session_idle_timeout="168h",
+    sso_session_idle_timeout="2h",
+    sso_session_max_lifespan="24h",
+    opts=resource_options,
+)
+
+required_action_configure_otp = keycloak.RequiredAction(
```
Comment (blarghmatey) [inline src/ol_infrastructure/substructure/keycloak/__main__.py:217]: Is this setting MFA as a required step for users? I'm not sure that we want to force that on everyone. I think it's a good idea, but we might want to give folks an option of not configuring it.

Comment (blarghmatey) [reply src/ol_infrastructure/substructure/keycloak/__main__.py:217]: Does that allow for the use of TOTP but not require it?

Comment (shaidar) [reply src/ol_infrastructure/substructure/keycloak/__main__.py:217]: So new users won't be forced to enable TOTP, but they do have the option to go to their account settings and add an authenticator app.

Comment (blarghmatey) [reply src/ol_infrastructure/substructure/keycloak/__main__.py:217]: Sounds good, thanks :+1:

**Review by blarghmatey — APPROVED** [review]

**Pull request merged**